### PR TITLE
build: bump toolchain to 2024-06-12

### PR DESCRIPTION
**Pull request by xxchan**
Signed-off-by: xxchan <xxchan22f@gmail.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- We cannot bump to > `2024-06-12`. There's a large change on TAIT: https://github.com/rust-lang/rust/pull/113169 https://github.com/rust-lang/rust/issues/126387, basically we will now have to define opaque types in its own module. (I've fixed some, but there are too many) In the future, there will be a `#[defines]` macro to ease this issue.
- We cannot use >= 06-07, <= 06-29 due to ICE https://github.com/rust-lang/rust/issues/127030

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
The deno's illegal instruction issue still persists :)

**Comment by xxchan** [comment]
We can merge this after 1.11 branches off.

**Comment by xxchan** [comment]
sim build ICE due to https://github.com/rust-lang/rust/issues/127030

File: lints/src/lib.rs [hunk 1]
```diff
@@ -36,13 +35,19 @@ pub fn register_lints(_sess: &rustc_session::Session, lint_store: &mut rustc_lin
     // -- Begin lint registration --
 
     // Preparation steps.
-    lint_store.register_early_pass(|| {
-        Box::<utils::format_args_collector::FormatArgsCollector>::default()
+    let format_args_storage = clippy_utils::macros::FormatArgsStorage::default();
```
Comment (BugenZhao) [inline lints/src/lib.rs:38]: This looks much better than an obscure thread-local storage. 😄

File: lints/src/utils/format_args_collector.rs [hunk 2]
```diff
@@ -12,15 +12,12 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Copied from `https://github.com/rust-lang/rust-clippy/blob/8b0bf6423dfaf5545014db85fcba7bc745beed4c/clippy_lints/src/utils/format_args_collector.rs`
-//!
-//! Init `AST_FORMAT_ARGS` before running the late pass, so that we can call `find_format_args`.
+//! Copied from `https://github.com/Alexendoo/rust-clippy/blob/c187bff864234e869dabcb41d2336639e29e2511/clippy_lints/src/utils/format_args_collector.rs`
```
Comment (BugenZhao) [inline lints/src/utils/format_args_collector.rs:15]: Why copied from a fork? 👀

File: src/connector/src/sink/kafka.rs [hunk 3]
```diff
@@ -403,7 +403,16 @@ struct KafkaPayloadWriter<'a> {
     config: &'a KafkaConfig,
 }
 
-pub type KafkaSinkDeliveryFuture = impl TryFuture<Ok = (), Error = SinkError> + Unpin + 'static;
+mod opaque_type {
```
Comment (BugenZhao) [inline src/connector/src/sink/kafka.rs:406]: Is this still necessary as we won't bump to `> 2024-06-12`?

**Review by BugenZhao — APPROVED** [review]

File: src/connector/src/sink/kafka.rs (hunk 3 above)
Comment (xxchan) [reply src/connector/src/sink/kafka.rs:406]: Not necessary, but I think it doesn't hurt.

File: src/object_store/src/object/mod.rs [hunk 4]
```diff
@@ -12,6 +12,11 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![expect(
+    unexpected_cfgs,
+    reason = "feature(hdfs-backend) is banned https://github.com/risingwavelabs/risingwave/pull/7875"
+)]
+
```
Comment (hzxa21) [inline src/object_store/src/object/mod.rs:19]: Will this cause the whole `object/mod.rs` to be excluded from the build?

Comment (xxchan) [reply src/object_store/src/object/mod.rs:19]: 👀 what do you mean? It only configures lint here

Comment (hzxa21) [reply src/object_store/src/object/mod.rs:19]: For some reason, vscode warns me that the code in this file is not active. That is why I am wondering whether the file is excluded.

Comment (xxchan) [reply src/object_store/src/object/mod.rs:19]: Seems a RA bug and already fixed.
https://github.com/rust-lang/rust-analyzer/issues/17817

**Pull request merged**